kyo-kernel/shared/src/test/scala/kyo/kernel/ArrowEffectTest.scala

```diff
@@ -337,4 +337,210 @@ class ArrowEffectTest extends Test:
+                    )
+
+            end Delim
+
+            "multi shot" in run {
+                Delim.shift[Int, Int, Any] { k =>
+                    k(42).map(a => k(42 + 1).map(b => a + b))
+                }.map(_ * 10)
+                    .handle(Delim.run)
+                    .map { v =>
+                        assert(v == (42 * 10) + ((42 + 1) * 10))
+                    }
+            }
+
+            "multi shot with other effect" in run {
+                Delim.shift[Int, Int, TestEffect2] { k =>
+                    k(42).map(a => testEffect2("a").map(k).map(b => a + b))
+                }.map(_ * 10)
+                    .handle(
+                        Delim.run,
+                        ArrowEffect.handle(Tag[TestEffect2], _) {
+                            [C] => (input, cont) => cont(input.size)
+                        }
+                    ).map { v =>
+                        assert(v == (42 * 10) + ("a".size * 10))
+                    }
+            }
+
+            "multiple shift with different effect sets" in run {
+                Delim.shift[Int, Int, TestEffect2] { k =>
+                    k(42).map { r =>
+                        testEffect2("a").map(v => r + v)
+                    }
+                }.map(_ * 10).map { v1 =>
+                    Delim.shift[Int, Int, TestEffect1] { k =>
+                        k(42).map { r =>
+                            testEffect1(v1).map(s => r + s.length)
+                        }
+                    }
+                }
+                    .handle(
+                        Delim.run,
+                        ArrowEffect.handle(Tag[TestEffect1], _) {
+                            [C] => (input, cont) => cont(input.toString)
+                        },
+                        ArrowEffect.handle(Tag[TestEffect2], _) {
+                            [C] => (input, cont) => cont(input.size)
+                        }
+                    ).map { v =>
+                        assert(v == 46)
+                    }
+            }
+
+            "short circuiting" in run {
+                def test(numbers: List[Int], expected: Int) =
+                    Kyo.foldLeft(numbers)(0) { (acc, n) =>
+                        if n < 0 || n == 42 then
+                            Delim.shift[Int, Int, Any] { _ => -1 }
+                        else
+                            acc + n
+                    }
+                        .handle(Delim.run)
+                        .map { r =>
+                            assert(r == expected)
+                        }
+
+                for
+                    _ <- test(List(1, 2), 3)
+                    _ <- test(List(1, 2, -1), -1)
+                    _ <- test(List(1, 2, 42, 3), -1)
+                yield succeed
+                end for
+            }
+        }
+
+        "flow effect with dynamic tags" - {
+            sealed trait Flow[+In, -Out] extends ArrowEffect[Flow.Op[In, Out, *], Id]
```

Comment:
@hearnadam I took the opportunity to explore an effect that blends emit and poll

Reply:
@johnhungerford you might be interested in this